[Bug Skip]: New failed cases 2026-1-22

### 🐛 Describe the bug with skip template

Cases:
~~op_ut,third_party.torch-xpu-ops.test.xpu.test_modules_xpu.TestModuleXPU,test_errors_nn_NLLLoss_xpu_float32~~
~~op_ut,third_party.torch-xpu-ops.test.xpu.test_modules_xpu.TestModuleXPU,test_errors_nn_NLLLoss_xpu_float64~~
op_ut,third_party.torch-xpu-ops.test.xpu.test_foreach_xpu.TestForeachXPU,test_addcmul_alpha_one_fma_parity_xpu_float32
op_ut,third_party.torch-xpu-ops.test.xpu.test_foreach_xpu.TestForeachXPU,test_addcmul_alpha_one_fma_parity_xpu_float64

Errorlog:
1.AssertionError: RuntimeError not raised
```
_______________ TestModuleXPU.test_errors_nn_NLLLoss_xpu_float32 _______________
[gw1] linux -- Python 3.10.19 /tmp/xpu-tool/Python/3.10.19/x64/bin/python
Traceback (most recent call last):
  File "/__w/torch-xpu-ops/torch-xpu-ops/pytorch/third_party/torch-xpu-ops/test/xpu/../../../../test/test_modules.py", line 879, in test_errors
    with self.assertRaisesRegex(error_input.error_type, error_input.error_regex):
  File "/tmp/xpu-tool/Python/3.10.19/x64/lib/python3.10/unittest/case.py", line 226, in __exit__
    self._raiseFailure("{} not raised".format(exc_name))
  File "/tmp/xpu-tool/Python/3.10.19/x64/lib/python3.10/unittest/case.py", line 163, in _raiseFailure
    raise self.test_case.failureException(msg)
AssertionError: RuntimeError not raised

_______________ TestModuleXPU.test_errors_nn_NLLLoss_xpu_float64 _______________
[gw1] linux -- Python 3.10.19 /tmp/xpu-tool/Python/3.10.19/x64/bin/python
Traceback (most recent call last):
  File "/__w/torch-xpu-ops/torch-xpu-ops/pytorch/third_party/torch-xpu-ops/test/xpu/../../../../test/test_modules.py", line 879, in test_errors
    with self.assertRaisesRegex(error_input.error_type, error_input.error_regex):
  File "/tmp/xpu-tool/Python/3.10.19/x64/lib/python3.10/unittest/case.py", line 226, in __exit__
    self._raiseFailure("{} not raised".format(exc_name))
  File "/tmp/xpu-tool/Python/3.10.19/x64/lib/python3.10/unittest/case.py", line 163, in _raiseFailure
    raise self.test_case.failureException(msg)
AssertionError: RuntimeError not raised
```

2.AssertionError: Tensor-likes are not equal!
```
_________ TestForeachXPU.test_addcmul_alpha_one_fma_parity_xpu_float32 _________
[gw0] linux -- Python 3.10.19 /tmp/xpu-tool/Python/3.10.19/x64/bin/python
Traceback (most recent call last):
  File "/__w/torch-xpu-ops/torch-xpu-ops/pytorch/third_party/torch-xpu-ops/test/xpu/../../../../test/test_foreach.py", line 1454, in test_addcmul_alpha_one_fma_parity
    self.assertEqual(
  File "/tmp/xpu-tool/Python/3.10.19/x64/lib/python3.10/site-packages/torch/testing/_internal/common_utils.py", line 4358, in assertEqual
    raise error_metas.pop()[0].to_error(  # type: ignore[index]
AssertionError: Tensor-likes are not equal!

_________ TestForeachXPU.test_addcmul_alpha_one_fma_parity_xpu_float64 _________
[gw0] linux -- Python 3.10.19 /tmp/xpu-tool/Python/3.10.19/x64/bin/python
Traceback (most recent call last):
  File "/__w/torch-xpu-ops/torch-xpu-ops/pytorch/third_party/torch-xpu-ops/test/xpu/../../../../test/test_foreach.py", line 1454, in test_addcmul_alpha_one_fma_parity
    self.assertEqual(
  File "/tmp/xpu-tool/Python/3.10.19/x64/lib/python3.10/site-packages/torch/testing/_internal/common_utils.py", line 4358, in assertEqual
    raise error_metas.pop()[0].to_error(  # type: ignore[index]
AssertionError: Tensor-likes are not equal!
```

pytorch : 9cec803f792649d10fecfc3f79bb1e38028ae99b

### Versions

<details>
<summary>Detail</summary>
Collecting environment information...
PyTorch version: 2.11.0a0+git9cec803
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: Ubuntu 24.04.3 LTS (x86_64)
GCC version: (Ubuntu 13.3.0-6ubuntu2~24.04) 13.3.0
Clang version: Could not collect
CMake version: version 3.31.6
Libc version: glibc-2.39

Python version: 3.10.19 (main, Oct 10 2025, 01:01:37) [GCC 13.3.0] (64-bit runtime)
Python platform: Linux-5.15.0-164-generic-x86_64-with-glibc2.39
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
Is XPU available: True
XPU used to build PyTorch: 20250301
Intel GPU driver version:
* intel-opencl-icd:	25.18.33578.42-1146~24.04
* libze1:	1.24.0.0-1146~24.04
Intel GPU models onboard:
* Intel(R) Data Center GPU Max 1100
* Intel(R) Data Center GPU Max 1100
* Intel(R) Data Center GPU Max 1100
* Intel(R) Data Center GPU Max 1100
* Intel(R) Data Center GPU Max 1100
* Intel(R) Data Center GPU Max 1100
* Intel(R) Data Center GPU Max 1100
* Intel(R) Data Center GPU Max 1100
Intel GPU models detected:
* [0] _XpuDeviceProperties(name='Intel(R) Data Center GPU Max 1100', platform_name='Intel(R) oneAPI Unified Runtime over Level-Zero', type='gpu', device_id=0xBDA, uuid=8680da0b-2f00-0000-0f00-000000000000, driver_version='1.6.33578+42', total_memory=49136MB, max_compute_units=448, gpu_eu_count=448, gpu_subslice_count=56, max_work_group_size=1024, max_num_sub_groups=64, sub_group_sizes=[16 32], has_fp16=1, has_fp64=1, has_atomic64=1)
* [1] _XpuDeviceProperties(name='Intel(R) Data Center GPU Max 1100', platform_name='Intel(R) oneAPI Unified Runtime over Level-Zero', type='gpu', device_id=0xBDA, uuid=8680da0b-2f00-0000-1600-000000000000, driver_version='1.6.33578+42', total_memory=49136MB, max_compute_units=448, gpu_eu_count=448, gpu_subslice_count=56, max_work_group_size=1024, max_num_sub_groups=64, sub_group_sizes=[16 32], has_fp16=1, has_fp64=1, has_atomic64=1)
* [2] _XpuDeviceProperties(name='Intel(R) Data Center GPU Max 1100', platform_name='Intel(R) oneAPI Unified Runtime over Level-Zero', type='gpu', device_id=0xBDA, uuid=8680da0b-2f00-0000-1a00-000000000000, driver_version='1.6.33578+42', total_memory=49136MB, max_compute_units=448, gpu_eu_count=448, gpu_subslice_count=56, max_work_group_size=1024, max_num_sub_groups=64, sub_group_sizes=[16 32], has_fp16=1, has_fp64=1, has_atomic64=1)
* [3] _XpuDeviceProperties(name='Intel(R) Data Center GPU Max 1100', platform_name='Intel(R) oneAPI Unified Runtime over Level-Zero', type='gpu', device_id=0xBDA, uuid=8680da0b-2f00-0000-1e00-000000000000, driver_version='1.6.33578+42', total_memory=49136MB, max_compute_units=448, gpu_eu_count=448, gpu_subslice_count=56, max_work_group_size=1024, max_num_sub_groups=64, sub_group_sizes=[16 32], has_fp16=1, has_fp64=1, has_atomic64=1)
* [4] _XpuDeviceProperties(name='Intel(R) Data Center GPU Max 1100', platform_name='Intel(R) oneAPI Unified Runtime over Level-Zero', type='gpu', device_id=0xBDA, uuid=8680da0b-2f00-0000-af00-000000000000, driver_version='1.6.33578+42', total_memory=49136MB, max_compute_units=448, gpu_eu_count=448, gpu_subslice_count=56, max_work_group_size=1024, max_num_sub_groups=64, sub_group_sizes=[16 32], has_fp16=1, has_fp64=1, has_atomic64=1)
* [5] _XpuDeviceProperties(name='Intel(R) Data Center GPU Max 1100', platform_name='Intel(R) oneAPI Unified Runtime over Level-Zero', type='gpu', device_id=0xBDA, uuid=8680da0b-2f00-0000-b300-000000000000, driver_version='1.6.33578+42', total_memory=49136MB, max_compute_units=448, gpu_eu_count=448, gpu_subslice_count=56, max_work_group_size=1024, max_num_sub_groups=64, sub_group_sizes=[16 32], has_fp16=1, has_fp64=1, has_atomic64=1)
* [6] _XpuDeviceProperties(name='Intel(R) Data Center GPU Max 1100', platform_name='Intel(R) oneAPI Unified Runtime over Level-Zero', type='gpu', device_id=0xBDA, uuid=8680da0b-2f00-0000-bf00-000000000000, driver_version='1.6.33578+42', total_memory=49136MB, max_compute_units=448, gpu_eu_count=448, gpu_subslice_count=56, max_work_group_size=1024, max_num_sub_groups=64, sub_group_sizes=[16 32], has_fp16=1, has_fp64=1, has_atomic64=1)
* [7] _XpuDeviceProperties(name='Intel(R) Data Center GPU Max 1100', platform_name='Intel(R) oneAPI Unified Runtime over Level-Zero', type='gpu', device_id=0xBDA, uuid=8680da0b-2f00-0000-c300-000000000000, driver_version='1.6.33578+42', total_memory=49136MB, max_compute_units=448, gpu_eu_count=448, gpu_subslice_count=56, max_work_group_size=1024, max_num_sub_groups=64, sub_group_sizes=[16 32], has_fp16=1, has_fp64=1, has_atomic64=1)
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Caching allocator config: N/A

CPU:
Architecture:                            x86_64
CPU op-mode(s):                          32-bit, 64-bit
Address sizes:                           46 bits physical, 57 bits virtual
Byte Order:                              Little Endian
CPU(s):                                  192
On-line CPU(s) list:                     0-191
Vendor ID:                               GenuineIntel
Model name:                              Intel(R) Xeon(R) Platinum 8468V
CPU family:                              6
Model:                                   143
Thread(s) per core:                      2
Core(s) per socket:                      48
Socket(s):                               2
Stepping:                                8
Frequency boost:                         enabled
CPU(s) scaling MHz:                      155%
CPU max MHz:                             2401.0000
CPU min MHz:                             800.0000
BogoMIPS:                                4800.00
Flags:                                   fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 cat_l2 cdp_l3 invpcid_single intel_ppin cdp_l2 ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb intel_pt avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local split_lock_detect avx_vnni avx512_bf16 wbnoinvd dtherm ida arat pln pts avx512vbmi umip pku ospke waitpkg avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg tme avx512_vpopcntdq la57 rdpid bus_lock_detect cldemote movdiri movdir64b enqcmd fsrm md_clear serialize tsxldtrk pconfig arch_lbr amx_bf16 avx512_fp16 amx_tile amx_int8 flush_l1d arch_capabilities ibpb_exit_to_user
Virtualization:                          VT-x
L1d cache:                               4.5 MiB (96 instances)
L1i cache:                               3 MiB (96 instances)
L2 cache:                                192 MiB (96 instances)
L3 cache:                                195 MiB (2 instances)
NUMA node(s):                            2
NUMA node0 CPU(s):                       0-47,96-143
NUMA node1 CPU(s):                       48-95,144-191
Vulnerability Gather data sampling:      Not affected
Vulnerability Indirect target selection: Not affected
Vulnerability Itlb multihit:             Not affected
Vulnerability L1tf:                      Not affected
Vulnerability Mds:                       Not affected
Vulnerability Meltdown:                  Not affected
Vulnerability Mmio stale data:           Not affected
Vulnerability Reg file data sampling:    Not affected
Vulnerability Retbleed:                  Not affected
Vulnerability Spec rstack overflow:      Not affected
Vulnerability Spec store bypass:         Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:                Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:                Mitigation; Enhanced / Automatic IBRS; IBPB conditional; PBRSB-eIBRS SW sequence; BHI BHI_DIS_S
Vulnerability Srbds:                     Not affected
Vulnerability Tsa:                       Not affected
Vulnerability Tsx async abort:           Not affected
Vulnerability Vmscape:                   Mitigation; IBPB before exit to userspace

Versions of relevant libraries:
[pip3] dpcpp-cpp-rt==2025.3.1
[pip3] impi-rt==2021.17.0
[pip3] intel-cmplr-lib-rt==2025.3.1
[pip3] intel-cmplr-lib-ur==2025.3.1
[pip3] intel-cmplr-lic-rt==2025.3.1
[pip3] intel-opencl-rt==2025.3.1
[pip3] intel-openmp==2025.3.1
[pip3] intel-pti==0.15.0
[pip3] intel-sycl-rt==2025.3.1
[pip3] mkl==2025.3.0
[pip3] mypy==1.16.0
[pip3] mypy_extensions==1.1.0
[pip3] numpy==1.23.2
[pip3] oneccl==2021.17.1
[pip3] oneccl-devel==2021.17.1
[pip3] onemkl-license==2025.3.0
[pip3] onemkl-sycl-blas==2025.3.0
[pip3] onemkl-sycl-dft==2025.3.0
[pip3] onemkl-sycl-lapack==2025.3.0
[pip3] onemkl-sycl-rng==2025.3.0
[pip3] onemkl-sycl-sparse==2025.3.0
[pip3] onnx==1.20.0
[pip3] onnx-ir==0.1.12
[pip3] onnxscript==0.5.4
[pip3] optree==0.13.0
[pip3] pytorch-triton-xpu==3.6.0+git225cdbde
[pip3] tbb==2022.3.0
[pip3] tcmlib==1.4.1
[pip3] torch==2.11.0a0+git9cec803
[pip3] torchao==0.16.0.dev20260121+xpu
[pip3] torchaudio==2.11.0a0+e123269
[pip3] torchvision==0.25.0a0+1e53952
[pip3] umf==1.0.2
[conda] Could not collect
</details>
TODO List:

- [ ] Remove all WA in PR https://github.com/pytorch/pytorch/pull/169354 caused by this issue

--- a/src/ATen/native/xpu/sycl/DeviceAddCmulCdiv.h
+++ b/src/ATen/native/xpu/sycl/DeviceAddCmulCdiv.h
@@ -1,0 +1,55 @@
+/*
+ * Copyright 2020-2026 Intel Corporation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+#pragma once
+
+#include <sycl/sycl.hpp>
+#include <functional>
+#include <type_traits>
+
+namespace at::native::xpu {
+
+// Trait to detect multiply-like functors (std::multiplies<T> for any T).
+// This enables the FMA fast-path for any instantiation of std::multiplies,
+// not just std::multiplies<opmath_t>.
+template <typename Op>
+struct is_multiply_op : std::false_type {};
+
+template <typename T>
+struct is_multiply_op<std::multiplies<T>> : std::true_type {};
+
+template <typename Op>
+inline constexpr bool is_multiply_op_v = is_multiply_op<Op>::value;
+
+// Computes input + alpha * op(tensor1, tensor2).
+// Uses sycl::fma for real floating-point types to ensure consistent
+// FMA behavior matching CUDA's pointwise_op_impl.
+template <typename opmath_t, typename Op>
+inline opmath_t pointwise_op_impl(
+    opmath_t input,
+    opmath_t tensor1,
+    opmath_t tensor2,
+    opmath_t alpha,
+    Op op) {
+  if (alpha == opmath_t(1)) {
+    if constexpr (is_multiply_op_v<Op> && std::is_floating_point_v<opmath_t>) {
+      return sycl::fma(tensor1, tensor2, input);
+    } else {
+      return input + op(tensor1, tensor2);
+    }
+  }
+  if constexpr (std::is_floating_point_v<opmath_t>) {
+    return sycl::fma(alpha, op(tensor1, tensor2), input);
+  } else {
+    return input + alpha * op(tensor1, tensor2);
+  }
+}
+
+} // namespace at::native::xpu

--- a/src/ATen/native/xpu/sycl/ForeachFunctors.h
+++ b/src/ATen/native/xpu/sycl/ForeachFunctors.h
@@ -12,10 +12,12 @@
 #include <ATen/OpMathType.h>
 #include <ATen/native/ForeachUtils.h>
 
+#include <ATen/native/xpu/sycl/DeviceAddCmulCdiv.h>
 #include <ATen/native/xpu/sycl/MultiTensorApply.h>
 #include <ATen/native/xpu/sycl/Pow.h>
 
 namespace at::native::xpu {
+
 namespace {
 
 inline void increment_version(TensorList tensors) {
@@ -253,11 +255,12 @@ struct PointwiseOpScalarFunctor {
         load_store(r_args[2], args[2], 0, i_start);
 #pragma unroll
         for (int ii = 0; ii < kILP; ii++) {
-          r_args[0][ii] = static_cast<T>(
-              static_cast<opmath_t>(r_args[0][ii]) +
-              scalar *
-                  op(static_cast<opmath_t>(r_args[1][ii]),
-                     static_cast<opmath_t>(r_args[2][ii])));
+          r_args[0][ii] = static_cast<T>(pointwise_op_impl<opmath_t>(
+              static_cast<opmath_t>(r_args[0][ii]),
+              static_cast<opmath_t>(r_args[1][ii]),
+              static_cast<opmath_t>(r_args[2][ii]),
+              scalar,
+              op));
         }
         load_store(args[res_arg_index], r_args[0], i_start, 0);
       }
@@ -268,11 +271,12 @@ struct PointwiseOpScalarFunctor {
             r_args, args, i_start, chunk_size, n, item_idx, item_range);
 #pragma unroll
         for (int ii = 0; ii < kILP; ii++) {
-          r_args[0][ii] = static_cast<T>(
-              static_cast<opmath_t>(r_args[0][ii]) +
-              scalar *
-                  op(static_cast<opmath_t>(r_args[1][ii]),
-                     static_cast<opmath_t>(r_args[2][ii])));
+          r_args[0][ii] = static_cast<T>(pointwise_op_impl<opmath_t>(
+              static_cast<opmath_t>(r_args[0][ii]),
+              static_cast<opmath_t>(r_args[1][ii]),
+              static_cast<opmath_t>(r_args[2][ii]),
+              scalar,
+              op));
         }
         store_args(
             args[res_arg_index],
@@ -319,11 +323,12 @@ struct PointwiseOpScalarListFunctor {
         load_store(r_args[2], args[2], 0, i_start);
 #pragma unroll
         for (int ii = 0; ii < kILP; ii++) {
-          r_args[0][ii] = static_cast<T>(
-              static_cast<opmath_t>(r_args[0][ii]) +
-              scalar *
-                  op(static_cast<opmath_t>(r_args[1][ii]),
-                     static_cast<opmath_t>(r_args[2][ii])));
+          r_args[0][ii] = static_cast<T>(pointwise_op_impl<opmath_t>(
+              static_cast<opmath_t>(r_args[0][ii]),
+              static_cast<opmath_t>(r_args[1][ii]),
+              static_cast<opmath_t>(r_args[2][ii]),
+              scalar,
+              op));
         }
         load_store(args[res_arg_index], r_args[0], i_start, 0);
       }
@@ -334,11 +339,12 @@ struct PointwiseOpScalarListFunctor {
             r_args, args, i_start, chunk_size, n, item_idx, item_range);
 #pragma unroll
         for (int ii = 0; ii < kILP; ii++) {
-          r_args[0][ii] = static_cast<T>(
-              static_cast<opmath_t>(r_args[0][ii]) +
-              scalar *
-                  op(static_cast<opmath_t>(r_args[1][ii]),
-                     static_cast<opmath_t>(r_args[2][ii])));
+          r_args[0][ii] = static_cast<T>(pointwise_op_impl<opmath_t>(
+              static_cast<opmath_t>(r_args[0][ii]),
+              static_cast<opmath_t>(r_args[1][ii]),
+              static_cast<opmath_t>(r_args[2][ii]),
+              scalar,
+              op));
         }
         store_args(
             args[res_arg_index],

--- a/src/ATen/native/xpu/sycl/PointwiseOpsKernels.cpp
+++ b/src/ATen/native/xpu/sycl/PointwiseOpsKernels.cpp
@@ -13,8 +13,8 @@
 #include <ATen/native/TensorIterator.h>
 #include <comm/xpu_aten.h>
 
+#include <ATen/native/xpu/sycl/DeviceAddCmulCdiv.h>
 #include <ATen/native/xpu/sycl/Loops.h>
-
 #include <ATen/native/xpu/sycl/PointwiseOpsKernels.h>
 
 namespace at::native::xpu {
@@ -23,8 +23,11 @@ template <typename scalar_t>
 struct AddcmulFunctor {
   using accscalar_t = at::acc_type_device<scalar_t, kXPU>;
   scalar_t operator()(scalar_t a, scalar_t b, scalar_t c) const {
-    return static_cast<accscalar_t>(a) +
-        alpha_ * static_cast<accscalar_t>(b) * static_cast<accscalar_t>(c);
+    auto input = static_cast<accscalar_t>(a);
+    auto t1 = static_cast<accscalar_t>(b);
+    auto t2 = static_cast<accscalar_t>(c);
+    return static_cast<scalar_t>(pointwise_op_impl(
+        input, t1, t2, alpha_, std::multiplies<accscalar_t>{}));
   }
 
   AddcmulFunctor(accscalar_t alpha) : alpha_(alpha) {}

